kasa/tapo/tapodevice.py

```diff
@@ -28,38 +30,71 @@ def __init__(
             transport=AesTransport(config=config or DeviceConfig(host=host)),
         )
         super().__init__(host=host, config=config, protocol=_protocol)
-        self._components: Optional[Dict[str, Any]] = None
+        self._components_raw: Optional[Dict[str, Any]] = None
+        self._components: Dict[str, int]
         self._state_information: Dict[str, Any] = {}
         self._discovery_info: Optional[Dict[str, Any]] = None
+        self.modules: Dict[str, Any] = {}
 
     async def update(self, update_children: bool = True):
         """Update the device."""
         if self.credentials is None or self.credentials.username is None:
             raise AuthenticationException("Tapo plug requires authentication.")
 
-        if self._components is None:
+        extra_reqs = {}
+        if self._components_raw is None:
             resp = await self.protocol.query("component_nego")
-            self._components = resp["component_nego"]
+            self._components_raw = resp["component_nego"]
+            self._components = {
+                comp["id"]: comp["ver_code"]
+                for comp in self._components_raw["component_list"]
+            }
+            await self._initialize_modules()
+
+            if "energy_monitoring" in self._components:
+                extra_reqs = {**extra_reqs,
+                    "get_energy_usage": None,
+                    "get_current_power": None,
+                }
 
         req = {
             "get_device_info": None,
             "get_device_usage": None,
             "get_device_time": None,
+            **extra_reqs,
         }
+
         resp = await self.protocol.query(req)
+        
         self._info = resp["get_device_info"]
         self._usage = resp["get_device_usage"]
         self._time = resp["get_device_time"]
+        # Emeter is not always available, but we set them still for now.
+        self._energy = resp.get("get_energy_usage")
+        self._emeter = resp.get("get_current_power")
 
         self._last_update = self._data = {
-            "components": self._components,
+            "components": self._components_raw,
             "info": self._info,
             "usage": self._usage,
             "time": self._time,
+            "energy": self._energy,
+            "emeter": self._emeter
         }
 
         _LOGGER.debug("Got an update: %s", self._data)
 
+    async def _initialize_modules(self):
+        """Initialize modules based on component negotiation response."""
+        if "energy_monitoring" in self._components:
+            self.emeter_type = "emeter"
+            self.modules["emeter"] = Emeter(self, self.emeter_type)
```

Comment:
Maybe this should also initialize the "extra_requests" to be used by future updates? Now it's rather hacky as we are checking for "energy_monitoring" separately in many different places...

Reply:
Agreed as per previous comment.  So this could be updated to add:

```
if "device" in self._components:

if "time" in self._components:
```

Reply:
Agreed, but I think this can wait for a separate PR that introduces proper module support for tapodevice class. That PR should pull the `module.py` to be part of the common API (i.e., out from `kasa/modules`) which should provide a common API for all supported devices.